connector/src/chain_data.rs

```diff
@@ -133,30 +133,18 @@ impl ChainData {
             self.account_bytes_stored = 0;
 
             for (_, writes) in self.accounts.iter_mut() {
-                let newest_rooted_write = writes
-                    .iter()
-                    .rev()
-                    .find(|w| {
-                        w.slot <= self.newest_rooted_slot
-                            && self
-                                .slots
-                                .get(&w.slot)
-                                .map(|s| {
-                                    // sometimes we seem not to get notifications about slots
-                                    // getting rooted, hence assume non-uncle slots < newest_rooted_slot
-                                    // are rooted too
-                                    s.status == SlotStatus::Rooted
-                                        || s.chain == self.best_chain_slot
-                                })
-                                // preserved account writes for deleted slots <= newest_rooted_slot
-                                // are expected to be rooted
-                                .unwrap_or(true)
-                    })
-                    .map(|w| w.slot)
-                    // no rooted write found: produce no effect, since writes > newest_rooted_slot are retained anyway
-                    .unwrap_or(self.newest_rooted_slot + 1);
-                writes
-                    .retain(|w| w.slot == newest_rooted_write || w.slot > self.newest_rooted_slot);
+                let newest_rooted_write_slot = Self::newest_rooted_write(
+                    writes,
+                    self.newest_rooted_slot,
+                    self.best_chain_slot,
+                    &self.slots,
+                )
+                .map(|w| w.slot)
+                // no rooted write found: produce no effect, since writes > newest_rooted_slot are retained anyway
+                .unwrap_or(self.newest_rooted_slot + 1);
+                writes.retain(|w| {
```

Comment:
Yeah, one could imagine an account appearing in a fork and never being rooted. Then all writes would eventually be removed but the pubkey entry would be retained.

The account views would skip that pubkey, so it's not user visible. And since it's unrelated to this PR I'm skipping it.